backends/memcheck/ceed-memcheck-vector.c

```diff
@@ -174,44 +252,129 @@ static int CeedVectorRestoreArray_Memcheck(CeedVector vec) {
+
+  for (CeedSize i = 0; i < length; i++) {
+    if (fabs(impl->array_allocated[i]) > CEED_EPSILON) impl->array_allocated[i] = 1. / impl->array_allocated[i];
+  }
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Compute x = alpha x
+//------------------------------------------------------------------------------
+static int CeedVectorScale_Memcheck(CeedVector x, CeedScalar alpha) {
+  CeedSize             length;
+  CeedVector_Memcheck *impl;
+
+  CeedCallBackend(CeedVectorGetData(x, &impl));
+  CeedCallBackend(CeedVectorGetLength(x, &length));
+
+  for (CeedSize i = 0; i < length; i++) impl->array_allocated[i] *= alpha;
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Compute y = alpha x + y
+//------------------------------------------------------------------------------
+static int CeedVectorAXPY_Memcheck(CeedVector y, CeedScalar alpha, CeedVector x) {
+  CeedSize             length;
+  CeedVector_Memcheck *impl_x, *impl_y;
+
+  CeedCallBackend(CeedVectorGetData(x, &impl_x));
+  CeedCallBackend(CeedVectorGetData(y, &impl_y));
+  CeedCallBackend(CeedVectorGetLength(y, &length));
+
+  for (CeedSize i = 0; i < length; i++) impl_y->array_allocated[i] += alpha * impl_x->array_allocated[i];
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Compute y = alpha x + beta y
+//------------------------------------------------------------------------------
+static int CeedVectorAXPBY_Memcheck(CeedVector y, CeedScalar alpha, CeedScalar beta, CeedVector x) {
+  CeedSize             length;
+  CeedVector_Memcheck *impl_x, *impl_y;
+
+  CeedCallBackend(CeedVectorGetData(x, &impl_x));
+  CeedCallBackend(CeedVectorGetData(y, &impl_y));
+  CeedCallBackend(CeedVectorGetLength(y, &length));
+
+  for (CeedSize i = 0; i < length; i++) impl_y->array_allocated[i] = alpha * impl_x->array_allocated[i] + beta * impl_y->array_allocated[i];
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Compute the pointwise multiplication w = x .* y
+//------------------------------------------------------------------------------
+static int CeedVectorPointwiseMult_Memcheck(CeedVector w, CeedVector x, CeedVector y) {
+  CeedSize             length;
+  CeedVector_Memcheck *impl_x, *impl_y, *impl_w;
+
+  CeedCallBackend(CeedVectorGetData(x, &impl_x));
+  CeedCallBackend(CeedVectorGetData(y, &impl_y));
+  CeedCallBackend(CeedVectorGetData(w, &impl_w));
+  CeedCallBackend(CeedVectorGetLength(w, &length));
+
+  if (!impl_w->array_allocated) CeedCallBackend(CeedVectorSetArray_Memcheck(w, CEED_MEM_HOST, CEED_COPY_VALUES, NULL));
+  assert(impl_w->array_allocated);
+  for (CeedSize i = 0; i < length; i++) impl_w->array_allocated[i] = impl_x->array_allocated[i] * impl_y->array_allocated[i];
```

Comment:
Should we add a NaN check here? Same rationale, but for `0 * inf` and `0 * -inf`, etc.